purchase_sale_inter_company/models/stock_picking.py

```diff
@@ -32,122 +40,199 @@ def _compute_state(self):
+                for dest_move, qty_done in dest_move_qty_update_dict.items():
+                    dest_move.quantity_done = qty_done
+                ic_pick._action_done()
+            except Exception as e:
+                if dest_company.sync_picking_failure_action == "raise":
+                    raise
+                else:
+                    pick._notify_picking_problem(
+                        pick.sale_id.auto_purchase_order_id,
+                        additional_note=str(e),
+                    )
+
+        # sync lots for move lines on pickings
         for pick in self.filtered(
             lambda x: x.location_dest_id.usage == "customer"
         ).sudo():
             purchase = pick.sale_id.auto_purchase_order_id
             if not purchase:
                 continue
-            purchase.picking_ids.write({"intercompany_picking_id": pick.id})
-            if not pick.intercompany_picking_id and purchase.picking_ids[0]:
-                pick.write({"intercompany_picking_id": purchase.picking_ids[0]})
-            pick._action_done_intercompany_actions(purchase)
-        return super()._action_done()
-
-    def _action_done_intercompany_actions(self, purchase):
-        self.ensure_one()
-        try:
-            pick = self
-            for move in pick.move_lines:
-                move_lines = move.move_line_ids
-                po_move_lines = (
-                    move.sale_line_id.auto_purchase_line_id.move_ids.filtered(
-                        lambda x, ic_pick=pick.intercompany_picking_id: x.picking_id
-                        == ic_pick
-                    ).mapped("move_line_ids")
-                )
-                if len(move_lines) != len(po_move_lines):
-                    note = _(
-                        "Mismatch between move lines with the "
-                        "corresponding PO %s for assigning "
-                        "quantities and lots from %s for product %s"
-                    ) % (purchase.name, pick.name, move.product_id.name)
-                    self.activity_schedule(
-                        "mail.mail_activity_data_warning",
-                        fields.Date.today(),
-                        note=note,
-                        # Try to notify someone relevant
-                        user_id=(
-                            pick.sale_id.user_id.id
-                            or pick.sale_id.team_id.user_id.id
-                            or SUPERUSER_ID,
-                        ),
+            dest_company = purchase.company_id
+            if not dest_company.sync_picking:
+                continue
+            try:
+                if not purchase.picking_ids:
+                    raise UserError(_("PO does not exist or has no receipts"))
+                ic_user = dest_company.intercompany_sale_user_id
+                purchase.picking_ids.write({"intercompany_picking_id": pick.id})
+                ic_pick = pick.intercompany_picking_id.with_user(ic_user)
```

Comment:
We need to add here `with_company(dest_company)` for those cases where the destination user has access to more than one company.